File: iosApp/Kwabor/Detail/CatalogDetailTypedContent.swift
import Shared
import SwiftUI

struct CatalogDetailTypedContent: View {
    let content: CatalogDetailContentUiModel
    let strings: CatalogDetailStrings
    let commonStrings: KwaborStrings

    @ViewBuilder
    var body: some View {
        if let place = content as? CatalogDetailContentUiModelPlace {
            placeContent(place)
        } else if let lodging = content as? CatalogDetailContentUiModelLodging {
            lodgingContent(lodging)
        } else if let food = content as? CatalogDetailContentUiModelFood {
            foodContent(food)
        } else if let nightlife = content as? CatalogDetailContentUiModelNightlife {
            nightlifeContent(nightlife)
        } else if let guide = content as? CatalogDetailContentUiModelGuide {
            guideContent(guide)
        } else if let event = content as? CatalogDetailContentUiModelEvent {
            eventContent(event)
        }
    }

    private func placeContent(_ place: CatalogDetailContentUiModelPlace) -> some View {
        CatalogDetailSection(title: place.heading) {
            CatalogDetailFact(label: strings.placeCategory, value: place.placeCategoryLabel)
            if let feeNote = place.feeNote {
                CatalogDetailFact(label: strings.feeNote, value: feeNote)
            }
        }
    }

    private func lodgingContent(_ lodging: CatalogDetailContentUiModelLodging) -> some View {
        CatalogDetailSection(title: lodging.heading) {
            CatalogDetailFacts(facts: lodging.facts)
            if !lodging.roomTypes.isEmpty {
                CatalogDetailSubheading(title: strings.roomTypes)
                CatalogDetailPricedItems(
                    items: lodging.roomTypes,
                    freeLabel: commonStrings.free,
                    transactional: false
                )
            }
        }
    }

    private func foodContent(_ food: CatalogDetailContentUiModelFood) -> some View {
        CatalogDetailSection(title: food.heading) {
            CatalogDetailLabelGroup(title: strings.cuisines, labels: food.cuisines)
            CatalogDetailLabelGroup(title: strings.meals, labels: food.meals)
            Text(food.reservationLabel)
                .foregroundStyle(KwaborDesignTokens.ColorToken.ink950)
            Text(food.menuAvailable ? strings.menuAvailable : strings.menuUnavailable)
                .foregroundStyle(KwaborDesignTokens.ColorToken.ink950)
        }
    }

    private func nightlifeContent(_ nightlife: CatalogDetailContentUiModelNightlife) -> some View {
        CatalogDetailSection(title: nightlife.heading) {
            CatalogDetailFacts(facts: nightlife.facts)
        }
    }

    private func guideContent(_ guide: CatalogDetailContentUiModelGuide) -> some View {
        CatalogDetailSection(title: guide.heading) {
            CatalogDetailLabelGroup(title: strings.languages, labels: guide.languages)
            CatalogDetailLabelGroup(title: strings.zones, labels: guide.zones)
            CatalogDetailLabelGroup(title: strings.specialties, labels: guide.specialties)
            CatalogDetailFacts(facts: guide.facts)
            if let price = guide.indicativePrice {
                CatalogDetailSubheading(title: strings.indicativePrice)
                CatalogDetailPriceTag(
                    label: detailPriceLabel(price, freeLabel: commonStrings.free),
                    transactional: false
                )
            }
        }
    }

    private func eventContent(_ event: CatalogDetailContentUiModelEvent) -> some View {
        CatalogDetailSection(title: event.heading) {
            if event.isEnded {
                Text(strings.eventEnded)
                    .font(.subheadline.weight(.semibold))
                    .foregroundStyle(KwaborDesignTokens.ColorToken.ink950)
                    .padding(.horizontal, KwaborDesignTokens.Spacing.md)
                    .frame(minHeight: KwaborDesignTokens.Sizing.touchTarget)
                    .background(KwaborDesignTokens.ColorToken.ink100)
                    .clipShape(Capsule())
                    .accessibilityLabel(strings.eventEnded)
            }
            CatalogDetailFact(label: strings.startsAt, value: event.startsAtLabel)
            if let endsAt = event.endsAtLabel {
                CatalogDetailFact(label: strings.endsAt, value: endsAt)
            }
            CatalogDetailFact(label: strings.venue, value: event.venueLabel)
            CatalogDetailFact(label: strings.organizer, value: event.organizerLabel)
            if let capacity = event.capacityLabel {
                CatalogDetailFact(label: strings.capacity, value: capacity)
            }
            CatalogDetailTicketing(
                ticketing: event.ticketing,
                strings: strings,
                freeLabel: commonStrings.free
            )
        }
    }
}

private struct CatalogDetailFacts: View {
    let facts: [CatalogDetailFactUiModel]

    var body: some View {
        ForEach(Array(facts.enumerated()), id: \.offset) { _, fact in
            CatalogDetailFact(label: fact.label, value: fact.value)
        }
    }
}

private struct CatalogDetailFact: View {
    let label: String
    let value: String

    var body: some View {
        VStack(alignment: .leading, spacing: KwaborDesignTokens.Spacing.xs) {
            Text(label)
                .font(.subheadline.weight(.semibold))
                .foregroundStyle(KwaborDesignTokens.ColorToken.ink700)
            Text(value)
                .font(.body)
                .foregroundStyle(KwaborDesignTokens.ColorToken.ink950)
                .fixedSize(horizontal: false, vertical: true)
        }
        .accessibilityElement(children: .combine)
    }
}

private struct CatalogDetailSubheading: View {
    let title: String

    var body: some View {
        Text(title)
            .font(.headline)
            .foregroundStyle(KwaborDesignTokens.ColorToken.ink950)
            .accessibilityAddTraits(.isHeader)
    }
}

private struct CatalogDetailLabelGroup: View {
    let title: String
    let labels: [String]

    var body: some View {
        if !labels.isEmpty {
            CatalogDetailSubheading(title: title)
            ScrollView(.horizontal) {
                LazyHStack(spacing: KwaborDesignTokens.Spacing.sm) {
                    ForEach(Array(labels.enumerated()), id: \.offset) { _, label in
                        Text(label)
                            .font(.subheadline)
                            .padding(.horizontal, KwaborDesignTokens.Spacing.md)
                            .frame(minHeight: KwaborDesignTokens.Sizing.touchTarget)
                            .background(KwaborDesignTokens.ColorToken.ink100)
                            .clipShape(Capsule())
                    }
                }
            }
            .scrollIndicators(.hidden)
        }
    }
}

private struct CatalogDetailPricedItems: View {
    let items: [CatalogDetailPricedItemUiModel]
    let freeLabel: String
    let transactional: Bool

    var body: some View {
        VStack(alignment: .leading, spacing: KwaborDesignTokens.Spacing.sm) {
            ForEach(Array(items.enumerated()), id: \.offset) { _, item in
                ViewThatFits(in: .horizontal) {
                    HStack(alignment: .firstTextBaseline) {
                        Text(item.label)
                        Spacer(minLength: KwaborDesignTokens.Spacing.md)
                        price(item.price)
                    }
                    VStack(alignment: .leading, spacing: KwaborDesignTokens.Spacing.xs) {
                        Text(item.label)
                        price(item.price)
                    }
                }
                .accessibilityElement(children: .combine)
            }
        }
    }

    private func price(_ amount: MoneyXof) -> some View {
        CatalogDetailPriceTag(
            label: detailPriceLabel(amount, freeLabel: freeLabel),
            transactional: transactional
        )
    }
}

private struct CatalogDetailTicketing: View {
    let ticketing: CatalogDetailTicketingUiModel
    let strings: CatalogDetailStrings
    let freeLabel: String

    @ViewBuilder
    var body: some View {
        CatalogDetailSubheading(title: strings.ticketing)
        if let free = ticketing as? CatalogDetailTicketingUiModelFree {
            Text(strings.freeEvent)
                .font(.body.weight(.semibold))
            Text(
                free.registrationAvailable
                    ? strings.registrationAvailable
                    : strings.registrationUnavailable
            )
            .foregroundStyle(KwaborDesignTokens.ColorToken.ink700)
        } else if let paid = ticketing as? CatalogDetailTicketingUiModelPaid {
            Text(strings.paidEvent)
                .font(.body.weight(.semibold))
            CatalogDetailPricedItems(
                items: paid.tiers,
                freeLabel: freeLabel,
                transactional: true
            )
        }
    }
}

File: androidApp/src/main/kotlin/com/kwabor/android/ui/screens/detail/CatalogDetailTypedContent.kt
package com.kwabor.android.ui.screens.detail

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import com.kwabor.android.R
import com.kwabor.android.design.KwaborRadius
import com.kwabor.android.design.KwaborSpacing
import com.kwabor.android.ui.components.PriceTag
import com.kwabor.android.ui.components.PriceTagMode
import com.kwabor.android.ui.components.PriceTagOptions
import com.kwabor.shared.i18n.CatalogDetailStrings
import com.kwabor.shared.i18n.KwaborStrings
import com.kwabor.shared.presentation.detail.CatalogDetailContentUiModel
import com.kwabor.shared.presentation.detail.CatalogDetailFactUiModel
import com.kwabor.shared.presentation.detail.CatalogDetailPricedItemUiModel
import com.kwabor.shared.presentation.detail.CatalogDetailTicketingUiModel

@Composable
internal fun DetailTypedContent(
    content: CatalogDetailContentUiModel,
    strings: CatalogDetailStrings,
    commonStrings: KwaborStrings,
    modifier: Modifier,
) {
    when (content) {
        is CatalogDetailContentUiModel.Place -> DetailPlace(content, strings, modifier)
        is CatalogDetailContentUiModel.Lodging -> DetailLodging(content, strings, commonStrings, modifier)
        is CatalogDetailContentUiModel.Food -> DetailFood(content, strings, modifier)
        is CatalogDetailContentUiModel.Nightlife -> DetailFacts(content.heading, content.facts, modifier)
        is CatalogDetailContentUiModel.Guide -> DetailGuide(content, strings, commonStrings, modifier)
        is CatalogDetailContentUiModel.Event -> DetailEvent(content, strings, commonStrings, modifier)
    }
}

@Composable
private fun DetailPlace(
    content: CatalogDetailContentUiModel.Place,
    strings: CatalogDetailStrings,
    modifier: Modifier,
) {
    TypedDetailSection(title = content.heading, modifier = modifier) {
        TypedDetailFact(label = strings.placeCategory, value = content.placeCategoryLabel)
        content.feeNote?.let { note -> TypedDetailFact(label = strings.feeNote, value = note) }
    }
}

@Composable
private fun DetailLodging(
    content: CatalogDetailContentUiModel.Lodging,
    strings: CatalogDetailStrings,
    commonStrings: KwaborStrings,
    modifier: Modifier,
) {
    TypedDetailSection(title = content.heading, modifier = modifier) {
        DetailFactList(content.facts)
        if (content.roomTypes.isNotEmpty()) {
            TypedDetailSubheading(strings.roomTypes)
            DetailPricedItems(content.roomTypes, commonStrings)
        }
    }
}

@Composable
private fun DetailFood(content: CatalogDetailContentUiModel.Food, strings: CatalogDetailStrings, modifier: Modifier) {
    TypedDetailSection(title = content.heading, modifier = modifier) {
        TypedDetailLabelGroup(strings.cuisines, content.cuisines)
        TypedDetailLabelGroup(strings.meals, content.meals)
        Text(text = content.reservationLabel, style = MaterialTheme.typography.bodyMedium)
        Text(
            text = if (content.menuAvailable) {
                strings.menuAvailable
            } else {
                stringResource(
                    R.string.detail_menu_unavailable,
                )
            },
            style = MaterialTheme.typography.bodyMedium,
        )
    }
}

@Composable
private fun DetailGuide(
    content: CatalogDetailContentUiModel.Guide,
    strings: CatalogDetailStrings,
    commonStrings: KwaborStrings,
    modifier: Modifier,
) {
    TypedDetailSection(title = content.heading, modifier = modifier) {
        TypedDetailLabelGroup(strings.languages, content.languages)
        TypedDetailLabelGroup(strings.zones, content.zones)
        TypedDetailLabelGroup(strings.specialties, content.specialties)
        DetailFactList(content.facts)
        content.indicativePrice?.let { price ->
            TypedDetailSubheading(strings.indicativePrice)
            PriceTag(
                price = price,
                strings = commonStrings,
                options = PriceTagOptions(mode = PriceTagMode.Full),
            )
        }
    }
}

@Composable
private fun DetailEvent(
    content: CatalogDetailContentUiModel.Event,
    strings: CatalogDetailStrings,
    commonStrings: KwaborStrings,
    modifier: Modifier,
) {
    TypedDetailSection(title = content.heading, modifier = modifier) {
        if (content.isEnded) DetailEventEndedBadge(strings.eventEnded)
        TypedDetailFact(label = strings.startsAt, value = content.startsAtLabel)
        content.endsAtLabel?.let { end -> TypedDetailFact(label = strings.endsAt, value = end) }
        TypedDetailFact(label = strings.venue, value = content.venueLabel)
        TypedDetailFact(label = strings.organizer, value = content.organizerLabel)
        content.capacityLabel?.let { capacity -> TypedDetailFact(label = strings.capacity, value = capacity) }
        TypedDetailSubheading(strings.ticketing)
        DetailEventTicketing(content.ticketing, strings, commonStrings)
    }
}

@Composable
private fun DetailEventEndedBadge(label: String) {
    Surface(
        shape = RoundedCornerShape(KwaborRadius.Pill),
        color = MaterialTheme.colorScheme.surfaceVariant,
    ) {
        Text(
            text = label,
            modifier = Modifier.padding(horizontal = KwaborSpacing.Md, vertical = KwaborSpacing.Sm),
            style = MaterialTheme.typography.labelLarge,
            fontWeight = FontWeight.SemiBold,
        )
    }
}

@Composable
private fun DetailEventTicketing(
    ticketing: CatalogDetailTicketingUiModel,
    strings: CatalogDetailStrings,
    commonStrings: KwaborStrings,
) {
    when (ticketing) {
        is CatalogDetailTicketingUiModel.Free -> {
            Text(text = strings.freeEvent, style = MaterialTheme.typography.bodyMedium)
            Text(
                text = if (ticketing.registrationAvailable) {
                    stringResource(R.string.detail_registration_available)
                } else {
                    stringResource(R.string.detail_registration_unavailable)
                },
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.secondary,
            )
        }
        is CatalogDetailTicketingUiModel.Paid -> {
            Text(text = strings.paidEvent, style = MaterialTheme.typography.bodyMedium)
            DetailPricedItems(ticketing.tiers, commonStrings)
        }
    }
}

@Composable
private fun DetailFacts(heading: String, facts: List<CatalogDetailFactUiModel>, modifier: Modifier) {
    TypedDetailSection(title = heading, modifier = modifier) { DetailFactList(facts) }
}

@Composable
private fun DetailFactList(facts: List<CatalogDetailFactUiModel>) {
    facts.forEach { fact -> TypedDetailFact(label = fact.label, value = fact.value) }
}

@Composable
private fun DetailPricedItems(items: List<CatalogDetailPricedItemUiModel>, strings: KwaborStrings) {
    LazyRow(horizontalArrangement = Arrangement.spacedBy(KwaborSpacing.Lg)) {
        items(items = items, key = CatalogDetailPricedItemUiModel::label) { item ->
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(text = item.label, style = MaterialTheme.typography.bodyMedium)
                Spacer(modifier = Modifier.width(KwaborSpacing.Md))
                PriceTag(
                    price = item.price,
                    strings = strings,
                    options = PriceTagOptions(mode = PriceTagMode.Full, transactional = true),
                )
            }
        }
    }
}

@Composable
private fun TypedDetailLabelGroup(heading: String, labels: List<String>) {
    if (labels.isEmpty()) return
    TypedDetailSubheading(heading)
    LazyRow(horizontalArrangement = Arrangement.spacedBy(KwaborSpacing.Sm)) {
        itemsIndexed(items = labels, key = { index, _ -> index }) { _, label ->
            Surface(shape = RoundedCornerShape(KwaborRadius.Pill), color = MaterialTheme.colorScheme.surfaceVariant) {
                Text(
                    text = label,
                    modifier = Modifier.padding(horizontal = KwaborSpacing.Md, vertical = KwaborSpacing.Sm),
                    style = MaterialTheme.typography.labelMedium,
                )
            }
        }
    }
}

@Composable
private fun TypedDetailFact(label: String, value: String) {
    Column(verticalArrangement = Arrangement.spacedBy(KwaborSpacing.Xs)) {
        Text(text = label, style = MaterialTheme.typography.labelMedium, color = MaterialTheme.colorScheme.secondary)
        Text(text = value, style = MaterialTheme.typography.bodyLarge)
    }
}

@Composable
private fun TypedDetailSection(title: String, modifier: Modifier = Modifier, content: @Composable () -> Unit) {
    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(KwaborSpacing.Md)) {
        Text(
            text = title,
            modifier = Modifier.semantics { heading() },
            style = MaterialTheme.typography.titleMedium,
            fontWeight = FontWeight.SemiBold,
        )
        content()
    }
}

@Composable
private fun TypedDetailSubheading(title: String) {
    Text(
        text = title,
        modifier = Modifier.semantics { heading() },
        style = MaterialTheme.typography.labelLarge,
        fontWeight = FontWeight.SemiBold,
    )
}
